pkgs/top-level/unit-files.nix

```diff
@@ -0,0 +1,105 @@
+  baseDirectory = ../unit;
+
+  entries = readDir baseDirectory;
+  shards = attrNames entries;
+
+  # Return the name mappings in a single shard
+  # Type: String -> AttrsOf Path
+  namesForShard = shard:
+    if entries.${shard} != "directory" then
+      # Ignore non-directories. Technically only Readme.md is allowed as a file in the base directory, so we could alternatively:
+      # - Assume that Readme.md is the only file and change the condition to `shard == "Readme.md"` for a minor performance improvement.
+      #   This would however cause very poor error messages if there's other files.
+      # - Ensure that Readme.md is the only file, throwing a better error message if that's not the case.
+      #   However for good code architecture, all error messages relating to `pkgs/unit` should be centralized in (TODO: Link to validity checks)
+      # Additionally in either of those alternatives, we would have to duplicate the hardcoding of "Readme.md"
+      { }
+    else
+      mapAttrs
+        (name: _: baseDirectory + "/${shard}/${name}/package.nix")
+        (readDir (baseDirectory + "/${shard}"));
+
+  # Merge the name mappings of all shards between the indices start (inclusive) and end (exclusive)
+  # This ignores any potential duplicates for efficiency
+  # Invariant: start < end
+  # Type: Int -> Int -> AttrsOf Path
+  binaryMerge = start: end:
+    if end - start >= 2 then
+      # If there's more than 2 elements, split the range in two, recurse on each part and merge the result
+      # The binaryMerge invariant is satisfied because each half will have at least 1 element
+      binaryMerge start (start + (end - start) / 2)
+      // binaryMerge (start + (end - start) / 2) end
+    else
+      # Otherwise there will be exactly 1 element due to the invariant, in which case we just return it directly
+      namesForShard (elemAt shards start);
+
+  # We could get the same result with a simple
+  #
+  #   builtins.foldl' (acc: entry: acc // namesForShard entry) {} shards;
+  #
+  # but this has rather poor evaluation performance for a large number of shards.
+  # This is because `l // r` has complexity O(l + r) with l and r being the number attributes on the left and right side respectively.
+  # So e.g. if we have n shards each with m entries, the above fold would be evaluated like this:
+  #     /\     O((n-1)m + m) = O(nm)
+  #    ⋰  m    ... (n times)
+  #   /\       O(2m + m)     = O(3m)
+  #  /\ m      O(m + m)      = O(2m)
+  # m  m
+  #
+  # Giving a total complexity of O(n²m)
+  #
+  # To improve this we make use of the fact that `//` is associative, allowing us to rearrange evaluation as a binary tree:
+  #         / \         O(1 * (0.5nm + 0.5nm)) = O(nm)
+  #      ⋰       ⋱      ... (log2(n) times)
+  #    / \       / \    O(n / 4 * (2m + 2m))   = O(nm)
+  #  /\   /\   /\   /\  O(n / 2 * (m + m))     = O(nm)
+  # m  m m  m m  m m  m
+  #
+  # Giving a total complexity of only O(log2(n)nm)
```

Comment:
nit(mathematician): `O(mn log n)`

Reply:
I'm using `log` instead of `log2` now, but I think I like the original order more because it's more directly comparable to the $n^2m$ result above